mypy/checkexpr.py

```diff
@@ -1350,91 +1347,56 @@ def overload_erased_call_targets(self,
                 matches.append(typ)
         return matches
 
-    def union_overload_matches(self, callables: List[CallableType]) -> Optional[CallableType]:
-        """Accepts a list of overload signatures and attempts to combine them together into a
-        new CallableType consisting of the union of all of the given arguments and return types.
-
-        Returns None if it is not possible to combine the different callables together in a
-        sound manner.
-
-        Assumes all of the given callables have argument counts compatible with the caller.
+    def union_overload_result(self,
+                              plausible_targets: List[CallableType],
+                              args: List[Expression],
+                              arg_types: List[Type],
+                              arg_kinds: List[int],
+                              arg_names: Optional[Sequence[Optional[str]]],
+                              callable_name: Optional[str],
+                              object_type: Optional[Type],
+                              context: Context,
+                              arg_messages: Optional[MessageBuilder] = None,
+                              ) -> Optional[Tuple[Type, Type]]:
```

Comment:
Yes, you are right, it can be much worse. Here is some more detailed analysis.

The relevant numbers that we have here are:
* number of (potentially matching) overloads, `p`
* number of unions in args, `k`
* numbers of items in every  union `n_i, i = 1, ..., k`

The worst case scenario is when every check succeeds, in this case we have `p * n_1 * ... * n_k` inference attempts, assuming an average `n` items per union this is roughly `p * n ** k`. This might look very bad, but I have some intuition why in practice average/typical performance may be much better. The key is that in practice it is hard to write an overload that will match so many combinations, plus there is a `return None` that will fully "unwind" the union call stack on first failure. So for any realistic overload the number of union item combinations that _successfully_ match will be limited by number of overloads `p`, for example to match successfully two args with types `Union[A, B]` and `Union[C, D]` one will need four overloads. Thus we will _typically_ make total at most `p ** 2` inference attempts.

Actually, now thinking more about this, we can also bailout soon on success, i.e. after splitting the first union, before we split the next one, we can check, maybe it already matches, for example:
```python
@overload
def f(x: int, y: object, z: object) -> int: ...
@overload
def f(x: str, y: object, z: object) -> str: ...

x: Union[int, str]
f(x, x, x)
```
There is not need to split the second and third overload. This will need just few extra lines, but can be a decent performance gain.

An additional comment in the defence of this algorithm is its simplicity of implementation (especially as compared to the breadth of covered use cases). Anyway, if you still think it is dangerous it is easy to put a limit on the call stack depth.

Reply:
> The key is that in practice it is hard to write an overload that will match so many combinations

I'm not sure if it'll actually be that hard in practice, especially if the overload is intended to match a wide variety of different calls using a mixture of typevars, `Any` types, object types, `*args`, `**kwargs`...

For example, here's a program which calls a simplified version of `map`. It's a bit contrived, but I don't think it's too far off from the sorts of programs some users might write in practice:

```python
from typing import overload, Any, TypeVar, Iterable, List, Dict, Callable, Union

S = TypeVar('S')

@overload
def simple_map() -> None: ...
@overload
def simple_map(func: Callable[..., S], *iterables: Iterable[Any]) -> S: ...
def simple_map(*args): pass

def format_row(*entries: object) -> str: pass

class DateTime: pass
JsonBlob = Dict[str, Any]
Column = Union[List[str], List[int], List[bool], List[float], List[DateTime], List[JsonBlob]]

def print_custom_table() -> None:
    a: Column
    b: Column
    c: Column
    d: Column
    e: Column
    f: Column
    g: Column

    for row in simple_map(format_row, a, b, c, d, e, f):
        print(row)
```

If I try running `simple_map` with 6 columns (a through f), it takes the new algorithm roughly a minute and a half on my computer to type check this program. If I tack on the 7th column (g), it takes mypy significantly longer. (I don't have an exact time unfortunately: I got bored of waiting after 10-15 minutes and killed the thing.)

I think type-checking after each split would probably only aggravate the problem, unfortunately (though I do think it could help in other cases).

> An additional comment in the defence of this algorithm is its simplicity of implementation (especially as compared to the breadth of covered use cases). Anyway, if you still think it is dangerous it is easy to put a limit on the call stack depth.

I also like the simplicity of this algorithm. (It's also not obvious to me how we could robustly solve this problem in any other way, in any case.)

But yeah, I still think we should add a limit of some sort. Even if the pathological cases end up being rare, I think it'd be a poor user experience if mypy starts unexpectedly stalling in those edge cases.

Reply:
> I think type-checking after each split would probably only aggravate the problem, unfortunately (though I do think it could help in other cases).

It actually helped. I added it (simplistic WIP style, need to polish it) and your example passed in a second (even with eight(!) union args) also few other complex examples I tried passed instantly.

> But yeah, I still think we should add a limit of some sort.

I don't want to set a calculated limit, I would add a nesting/split level limit probably 5, and then raise `TooManyUnionsError`, so that we can better diagnose an error and add a note like:
```
main.py: error: No overload variant blah-blah-blah..
main.py: note: Not all union combinations were tried because there are too many unions
```

There are some ideas to avoid hitting the limit, like add some back-tracking add multi-seed (like start splitting the unions from last arg as well). But I will not do this, because this will only add code complexity while gaining some super-rare corner cases.